### PR TITLE
Deeplake vectore store

### PR DESCRIPTION
**Pull request by adolkhan**
Added support for deeplake Vector store. Also provided Jupyter notebook with example on how to use deeplake with Llama index

### CODE REVIEW

**Comment by jerryjliu** [comment]
sweet! will take a look soon :)

File: gpt_index/indices/vector_store/vector_indices.py [hunk 1]
```diff
@@ -435,6 +435,73 @@ def __init__(
         )
 
 
+class GPTDeepLakeIndex(GPTVectorStoreIndex):
+    """GPT DeepLake Vector Store.
+
+    In this vector store we store the text, its embedding and
+    a few pieces of its metadata in a deeplake dataset. This implemnetation
```
Comment (jerryjliu) [inline gpt_index/indices/vector_store/vector_indices.py:442]: ubernit: implementation

File: gpt_index/vector_stores/deeplake.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        with self.ds:
+            for id in sorted(ids)[::-1]:
+                self.ds.pop(id)
+
+            self.ds.commit(f"deleted {len(ids)} samples", allow_empty=True)
+            self.ds.summary()
+        return True
+
+    def query(self, query: VectorStoreQuery) -> VectorStoreQueryResult:
```
Comment (jerryjliu) [inline gpt_index/vector_stores/deeplake.py:228]: add docstring

File: gpt_index/vector_stores/deeplake.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            "ingestion_batch_size": self.ingestion_batch_size,
+            "ingestion_num_workers": self.num_workers,
+        }
+
+    @property
+    def client(self) -> None:
+        """Get client."""
+        return self.ds
+
+    def add(self, embedding_results: List[NodeEmbeddingResult]) -> List[str]:
```
Comment (jerryjliu) [inline gpt_index/vector_stores/deeplake.py:162]: also add docstring (otherwise linting fails)

**Review by jerryjliu — COMMENTED** [review]
this is an awesome start.

to get this to work as part of a composed graph you may need to add the following changes (we will refactor this later): 
- subclass IndexDict in gpt_index.data_structs.data_structs_v2 and add a DeepLakeIndexDict
- add as a type in gpt_index.data_structs.struct_type
- add to `INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS` and `INDEX_STRUCT_TYPE_TO_INDEX_CLASS` in gpt_index.indices.registry 

also, outside of this PR, any interest in also adding a loader to llama-hub? allows users to load in existing data in DeepLake if they want to use our constructs on top of it

**Comment by adolkhan** [comment]
Thx a lot!, will address your comments soon

File: gpt_index/indices/vector_store/vector_indices.py (hunk 1 above)
Comment (jerryjliu) [reply gpt_index/indices/vector_store/vector_indices.py:442]: oh it just means a nit but like a super small nit :)

File: gpt_index/vector_stores/deeplake.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+from gpt_index.data_structs.node_v2 import DocumentRelationship, Node
+from gpt_index.indices.query.embedding_utils import get_top_k_embeddings
+from gpt_index.vector_stores.types import (
+    NodeEmbeddingResult,
+    VectorStore,
+    VectorStoreQuery,
+    VectorStoreQueryMode,
+    VectorStoreQueryResult,
+)
+from torch.utils.data import DataLoader, Dataset
```
Comment (jerryjliu) [inline gpt_index/vector_stores/deeplake.py:21]: this would probably also need to be a lazy import. no worries, i can help fix

Comment (jerryjliu) [reply gpt_index/vector_stores/deeplake.py:21]: also, it doesn't seem like it's needed here

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**